--!strict

------------------------------------------------------------------------------------------------------------------------------------------

local ReplicatedStorage = game:GetService("ReplicatedStorage")

------------------------------------------------------------------------------------------------------------------------------------------

local luau_class = ReplicatedStorage["luau-class"]

local list_test = require(luau_class.tests.collection_test.list)
local queue_test = require(luau_class.tests.collection_test.queue)

------------------------------------------------------------------------------------------------------------------------------------------

local function main()
    queue_test.test()
end

------------------------------------------------------------------------------------------------------------------------------------------

main()

------------------------------------------------------------------------------------------------------------------------------------------