--!strict

---------------------------------------------------------------------------------------------------------------------------

local module = {}

---------------------------------------------------------------------------------------------------------------------------

local format = string.format

---------------------------------------------------------------------------------------------------------------------------

module.Inherit = function(new_str: string, old_str: string): string
    return format("%s.%s", old_str, new_str)
end

module.GetClassName = function(str: string): string
    local last_dot = string.find(str, "%.[^%.]*$")
    if last_dot then
        return string.sub(str, last_dot + 1)
    end
    return str
end

---------------------------------------------------------------------------------------------------------------------------

return module

---------------------------------------------------------------------------------------------------------------------------